plugins/lua/blueprint.lua

```diff
@@ -1,14 +1,161 @@
 local _ENV = mkmodule('plugins.blueprint')
 
---[[
+local utils = require('utils')
 
- Native functions:
+-- the info here is very basic and minimal, so hopefully we won't need to change
+-- it when features are added and the full blueprint docs in Plugins.rst are
+-- updated.
+local help_text = [=[
 
- * dig(start, end, name)
- * build(start, end, name)
- * place(start, end, name)
- * query(start, end, name)
+blueprint
+=========
 
---]]
+Records the structure of a portion of your fortress in quickfort blueprints.
+
+Usage:
+
+    blueprint <width> <height> [<depth>] [<name> [<phases>]] [<options>]
+    blueprint gui [<name> [<phases>]] [<options>]
+
+Examples:
+
+blueprint gui
+    Runs gui/blueprint, the interactive blueprint frontend, where all
+    configuration can be set visually and interactively.
+
+blueprint 30 40 bedrooms
+    Generates blueprints for an area 30 tiles wide by 40 tiles tall, starting
+    from the active cursor on the current z-level. Output files are written to
+    the "blueprints" directory.
+
+See the online DFHack documentation for more examples and details.
+]=]
+
+function print_help() print(help_text) end
+
+valid_phases = utils.invert{
+    'dig',
+    'build',
+    'place',
+    'query',
+}
+
+local function parse_cursor(opts, arg)
+    local _, _, x, y, z = arg:find('^(%d+),(%d+),(%d+)$')
+    if not x then
+        qerror(('invalid argument for --cursor option: "%s"; expected format' ..
+                ' is "<x>,<y>,<z>", for example: "30,60,150"'):format(arg))
+    end
+    -- be careful not to replace struct members when called from C++, but also
+    -- create the table as needed when called from lua
+    if not opts.start then opts.start = {} end
+    opts.start.x = tonumber(x)
+    opts.start.y = tonumber(y)
+    opts.start.z = tonumber(z)
+end
+
+local function parse_positionals(opts, args, start_argidx)
+    local argidx = start_argidx or 1
+
+    -- set defaults
+    opts.name, opts.auto_phase = 'blueprint', true
+
+    local name = args[argidx]
+    if not name then return end
+    if name == '' then
+        qerror(('invalid basename: "%s"; must be a valid, non-empty pathname')
+               :format(args[argidx]))
+    end
+    argidx = argidx + 1
+    -- normalize paths to forward slashes
+    opts.name = name:gsub(package.config:sub(1,1), "/")
+
+    local auto_phase = true
+    local phase = args[argidx]
+    while phase do
+        if valid_phases[phase] then
```

Comment:
Should we do something if the argument isn't valid? I tried `blueprint 1 1 build foo` and it wrote four `build-*.csv` files, ignoring the `foo` argument. Both ``blueprint 1 1 build foo place` and ``blueprint 1 1 build place foo` write just `build-place.csv`. (I know I'm using `build` as the blueprint name, yeah... I wasn't actually confused, just wanted to see what would happen).

Reply:
this makes sense. I'm not sure why I decided to just ignore invalid phases.

done, and updated tests for the behavior change.